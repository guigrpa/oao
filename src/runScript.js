// @flow

import { removeAllListeners, addListener } from 'storyboard';
import parallelConsoleListener from 'storyboard-listener-console-parallel';

import { readAllSpecs } from './utils/readSpecs';
import { exec } from './utils/shell';

type Options = {
  src: string,
  ignoreSrc?: string,
  parallel?: boolean,
  parallelLogs?: boolean,
  ignoreErrors?: boolean,
};

const run = async (
  script: string,
  { src, ignoreSrc, parallel, parallelLogs, ignoreErrors }: Options
) => {
  if (parallel && parallelLogs) {
    removeAllListeners();
    addListener(parallelConsoleListener);
  }
  const allSpecs = await readAllSpecs(src, ignoreSrc, false);
  const allPromises = [];
  const pkgNames = Object.keys(allSpecs);
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    const { pkgPath, specs: prevSpecs } = allSpecs[pkgName];
    if (prevSpecs.scripts && prevSpecs.scripts[script]) {
      let promise = exec(`yarn run ${script}`, {
        cwd: pkgPath,
        bareLogs: parallelLogs,
      });
      if (ignoreErrors) promise = promise.catch(() => {});
      if (!parallel) {
        await promise;
      } else {
        allPromises.push(promise);
      }
    }
  }

  // If parallel logs are enabled, we have to manually exit
  if (parallel && parallelLogs) {
    await Promise.all(allPromises);
    process.exit(0);
  }
};

export default run;
