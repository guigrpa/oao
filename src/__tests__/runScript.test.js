/* eslint-env jest */
/* eslint-disable global-require */

import runScript from '../runScript';

jest.mock('../utils/shell');

describe('RUN-SCRIPT command', () => {
  it('executes the specified script on all sub-packages', async () => {
    const helpers = require('../utils/shell');
    await runScript('start', { src: 'test/fixtures/packages/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });
});
