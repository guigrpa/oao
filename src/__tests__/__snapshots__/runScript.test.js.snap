// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RUN-SCRIPT command executes the specified script on all sub-packages 1`] = `
Array [
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-b",
    },
  ],
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-c",
    },
  ],
]
`;
